src/mpi4py/util/pkl5.py

```diff
@@ -0,0 +1,534 @@
+        mrecv(next(rmsg), bigmpi(data))
+        for rbuf in bufs:
+            mrecv(next(rmsg), bigmpi(rbuf))
+    if status is not None:
+        status.Set_elements(MPI.BYTE, sum(info))
+    return (data, bufs)
+
+
+def _mrecv(message, status):
+    def mrecv(rmsg, buf):
+        MPI.Message.Recv(rmsg, buf)
+    data, bufs = _mrecv_data(message, mrecv, status)
+    return pickle.loads(data, bufs)
+
+
+def _imrecv(message):
+    rreqs = []
+    def mrecv(rmsg, buf):
+        rreqs.append(MPI.Message.Irecv(rmsg, buf))
+    data, bufs = _mrecv_data(message, mrecv)
+    request = Request(rreqs)
+    setattr(request, '_data_bufs', (data, bufs))
+    return request
+
+
+def _req_load(request):
+    data_bufs = getattr(request, '_data_bufs', None)
+    if request == MPI.REQUEST_NULL and data_bufs is not None:
+        delattr(request, '_data_bufs')
+    if data_bufs is not None:
+        data, bufs = data_bufs
+        obj = pickle.loads(data, bufs)
+        return obj
+    return None
+
+
+def _test(request, test, status):
+    statuses = None if status is None else [status]
+    flag = test(request, statuses)
+    if flag:
+        obj = _req_load(request)
+        return (flag, obj)
+    return (flag, None)
+
+
+def _testall(requests, testall, statuses):
+    if isinstance(statuses, list):
+        for _ in range(len(requests) - len(statuses)):
+            statuses.append(Status())
+    reqarray = []
+    stsarray = None
+    for req in requests:
+        reqarray.extend(req)
+    if statuses is not None:
+        stsarray = []
+        for req, sts in zip(requests, statuses):
+            stsarray.extend([sts] * len(req))
+    flag = testall(reqarray, stsarray)
+    if flag:
+        objs = [_req_load(req) for req in requests]
+        return (flag, objs)
+    return (flag, None)
+
+
+class Request(tuple):
+    """Request."""
+
+    def __new__(cls, request=None):
+        """Create and return a new object."""
+        if request is None:
+            request = (MPI.REQUEST_NULL,)
+        if isinstance(request, MPI.Request):
+            request = (request,)
+        return super().__new__(cls, request)
+
+    def __eq__(self, other):
+        """Return ``self==other``."""
+        if isinstance(other, Request):
+            return tuple(self) == tuple(other)
+        if isinstance(other, MPI.Request):
+            return all(req == other for req in self)
+        return NotImplemented
+
+    def __ne__(self, other):
+        """Return ``self!=other``."""
+        if isinstance(other, Request):
+            return tuple(self) != tuple(other)
+        if isinstance(other, MPI.Request):
+            return any(req != other for req in self)
+        return NotImplemented
+
+    def __bool__(self):
+        """Return ``bool(self)``."""
+        return any(req for req in self)
+
+    def Free(self) -> None:
+        """Free a communication request."""
+        # pylint: disable=invalid-name
+        for req in self:
+            req.Free()
```

Comment:
Can we define `Request.__del__` to call `Free` here so that it gets cleaned up when out of scope? At least you did this here (`del_Request` calls `MPI_Request_Free`, like `Free` does): https://github.com/mpi4py/mpi4py/blob/bb5289d1531329b5ea28780996b51399ddf72200/src/mpi4py/MPI/Request.pyx#L13-L15

Reply:
No, sorry, all Cython  `__dealloc__` methods in mpi4py do nothing, the `flags` slot is always zero. The implementation is half-way, that's the reason `del_Request()` is implemented but effectively unused. For now, until I figure out a robust way to improve things, users are responsible of deallocating every new MPI handle then create. Very unpythonic, but I'm not a fan of promising something I cannot deliver. Most MPI implementations use refcounting to manage MPI handles, unfortunately, that machinery is not exposed (and certainly not part of the MPI standard).

Reply:
OK, thanks for clarifying. 
> users are responsible of deallocating every new MPI handle then create

Perhaps we better mention this somewhere in the docs. Let me create an issue to track this.